### PR TITLE
Switch to netlink.RouteListFilteredIter() API

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Optimise RouteTable resync methods:
* Add benchmark for resync with various numbers of routes. 
* Switch to netlink.RouteListFilteredIter() API, which avoids a series of large slice allocations.
* Use benchmark to find more wins:

  * Use  deltatracker.WithValuesEqualFn, DeepEquals is the default and it was much slower.
  * Pre-allocate the seenKeys set, saves the expense of copying it multiple times as it grows.
  * Copy the netlink Route once per loop to a scratch variable to avoid escape to the heap.

Overall, significantly reduces allocations from when there are many routes to resync and significantly increases performance.

Benchmarks for resyncing 64k routes; 50% reduction in resync time; 80% saving in bytes allocated; GC-included CPU usage down by 60%.
```
v3.28:

BenchmarkResync65536-12    	       7	 156498055 ns/op	 242857143 ncpu/op	203556697 B/op	  790473 allocs/op

master:

BenchmarkResync65536-12    	       7	 143550698 ns/op	 270000000 ncpu/op	208147700 B/op	  724999 allocs/op

This PR: 

BenchmarkResync65536-12    	      13	  84051862 ns/op	 108461538 ncpu/op	36912350 B/op	  526080 allocs/op
```
## Related issues/PRs

Follow on from #8979 

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Felix's route resync logic has been optimised; it now uses 50% less CPU time and 80% less memory.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/routetable/bench_test.go
Comment (aaaaaaaalex) [inline felix/routetable/bench_test.go:101]: should there be error handling in the above RouteUpdate, such that the loop has an alternative break condition?
To me it seems this condition will always be false currently, unless a test is set up with more than 65535 `numRoutes`

File: felix/routetable/route_table.go
Comment (aaaaaaaalex) [inline felix/routetable/route_table.go:802]: :muscle:

**Review by aaaaaaaalex — APPROVED** [review]
Just one potential change, might be nothing

File: felix/routetable/bench_test.go
Comment (fasaxc) [reply felix/routetable/bench_test.go:101]: This is here to prevent someone from adding `BenchmarkResync1000000` without fixing this bit of the code to actually be capable of generating that many.

**Comment by fasaxc** [comment]
Just holding this back until I've got the RouteTable merge to Enterprise done.  Might want to upstream some of the fixes there and I don't want this to conflict...

**Pull request merged**
